src/app/AttributePathParams.h

```diff
@@ -19,48 +19,54 @@
+    // TODO: (#11420) This class is overlapped with ClusterInfo class, need to do a clean up.
+    AttributePathParams(EndpointId aEndpointId, ClusterId aClusterId) :
+        AttributePathParams(aEndpointId, aClusterId, ClusterInfo::kInvalidAttributeId, ClusterInfo::kInvalidListIndex)
+    {}
 
     AttributePathParams(EndpointId aEndpointId, ClusterId aClusterId, AttributeId aFieldId) :
-        AttributePathParams(0, aEndpointId, aClusterId, aFieldId, 0, chip::app::AttributePathParams::Flags::kFieldIdValid)
+        AttributePathParams(aEndpointId, aClusterId, aFieldId, ClusterInfo::kInvalidListIndex)
     {}
 
     AttributePathParams(EndpointId aEndpointId, ClusterId aClusterId, AttributeId aFieldId, ListIndex aListIndex) :
-        AttributePathParams(0, aEndpointId, aClusterId, aFieldId, aListIndex,
-                            BitFlags<Flags>(chip::app::AttributePathParams::Flags::kFieldIdValid,
-                                            chip::app::AttributePathParams::Flags::kListIndexValid))
-    {}
-
-    AttributePathParams(NodeId aNodeId, EndpointId aEndpointId, ClusterId aClusterId, AttributeId aFieldId, ListIndex aListIndex,
-                        const BitFlags<Flags> aFlags) :
-        mNodeId(aNodeId),
-        mEndpointId(aEndpointId), mClusterId(aClusterId), mFieldId(aFieldId), mListIndex(aListIndex), mFlags(aFlags)
+        mEndpointId(aEndpointId), mClusterId(aClusterId), mFieldId(aFieldId), mListIndex(aListIndex)
     {}
 
     AttributePathParams() {}
-    NodeId mNodeId         = 0;
-    EndpointId mEndpointId = 0;
-    ClusterId mClusterId   = 0;
-    AttributeId mFieldId   = 0;
-    ListIndex mListIndex   = 0;
-    BitFlags<Flags> mFlags;
+
+    CHIP_ERROR BuildAttributePath(AttributePathIB::Builder & aBuilder) const;
+
+    bool HasWildcard() const { return HasWildcardEndpointId() || HasWildcardClusterId() || HasWildcardAttributeId(); }
+
+    /**
+     * Check that the path meets some basic constraints of an attribute path: If list index is not wildcard, then field id must not
+     * be wildcard. This does not verify that the attribute being targeted is actually of list type when the list index is not
+     * wildcard.
+     */
+    bool IsValidAttributePath() const { return HasWildcardListIndex() || !HasWildcardAttributeId(); }
+
+    inline bool HasWildcardEndpointId() const { return mEndpointId == ClusterInfo::kInvalidEndpointId; }
+    inline bool HasWildcardClusterId() const { return mClusterId == ClusterInfo::kInvalidClusterId; }
+    inline bool HasWildcardAttributeId() const { return mFieldId == ClusterInfo::kInvalidAttributeId; }
+    inline bool HasWildcardListIndex() const { return mListIndex == ClusterInfo::kInvalidListIndex; }
+
+    EndpointId mEndpointId = ClusterInfo::kInvalidEndpointId;
+    ClusterId mClusterId   = ClusterInfo::kInvalidClusterId;
+    AttributeId mFieldId   = ClusterInfo::kInvalidAttributeId;
```

Comment:
nit: shall we use mAttributeId instead of mFieldId to avoid confusion?

Reply:
@yunhanw-google I guess you will do this in a PR with a simple find / replace?

Reply:
https://github.com/project-chip/connectedhomeip/pull/11514